runners/pvc_autoresizer.go

```diff
@@ -0,0 +1,255 @@
+package runners
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create
+
+const resizeEnableIndexKey = ".metadata.annotations[resize.topolvm.io/enabled]"
+const storageClassNameIndexKey = ".spec.storageClassName"
+
+// NewPVCAutoresizer returns a new pvcAutoresizer struct
+func NewPVCAutoresizer(mc MetricsClient, interval time.Duration, recorder record.EventRecorder) *PVCAutoresizer {
+
+	return &PVCAutoresizer{
+		metricsClient: mc,
+		interval:      interval,
+		recorder:      recorder,
+	}
+}
+
+// InjectClient implements inject.Client
+func (w *PVCAutoresizer) InjectClient(c client.Client) error {
+	w.client = c
+	return nil
+}
+
+// InjectLogger implements inject.Logger
+func (w *PVCAutoresizer) InjectLogger(log logr.Logger) error {
+	w.log = log
+	return nil
+}
+
+// PVCAutoresizer is a runner which resize PVC capacity automatically
+type PVCAutoresizer struct {
+	client        client.Client
+	metricsClient MetricsClient
+	interval      time.Duration
+	log           logr.Logger
+	recorder      record.EventRecorder
+}
+
+// Start implements manager.Runnable
+func (w *PVCAutoresizer) Start(ch <-chan struct{}) error {
+	ticker := time.NewTicker(w.interval)
+	ctx := context.Background()
+
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ch:
+			return nil
+		case <-ticker.C:
+			err := w.notifyPVCEvent(ctx)
```

Comment:
This method name looks strange because what it does
is to check and resize PVCs.